nixos/modules/services/networking/teleport.nix

```
@@ -0,0 +1,88 @@
+{ config, pkgs, lib, ... }:
+
+with lib;
+
+let
+  cfg = config.services.teleport;
+  pidFilePath = "/run/teleport.pid";
+  configJSON = pkgs.writeText "teleport.json" (builtins.toJSON cfg.settings);
+  cmdlineArgs = [ "--pid-file ${pidFilePath}" ]
+    ++ optional cfg.debug.enable "--debug"
+    ++ optional cfg.diag.enable "--diag-addr=${cfg.diag.addr}:${toString cfg.diag.port}"
+    ++ optional (cfg.settings != { }) "--config=${configJSON}"
+    ++ cfg.extraOpts;
+in
+{
+  options = {
+    services.teleport = with lib.types; {
+      enable = mkEnableOption "Teleport service.";
+
+      settings = mkOption {
+        type = attrs;
+        default = { };
+        example = literalExpression ''
+          {
+            teleport = {
+              nodename = "client";
+              advertise_ip = "192.168.1.2";
+              auth_token = "60bdc117-8ff4-478d-95e4-9914597847eb";
+              auth_servers = [ "192.168.1.1:3025" ];
+          }
+        '';
+        description = "Reference: https://goteleport.com/docs/setup/reference/config/";
+      };
+
+      debug.enable = mkEnableOption "debug.";
+
+      diag = {
+        enable = mkEnableOption "endpoints for monitoring purposes. Details: https://goteleport.com/docs/setup/admin/troubleshooting/#troubleshooting";
+
+        addr = mkOption {
+          type = str;
+          default = "127.0.0.1";
+          description = "Metrics and diagnostics address.";
+        };
+
+        port = mkOption {
+          type = int;
+          default = 3000;
+          description = "Metrics and diagnostics port.";
+        };
+      };
+
+      extraOpts = mkOption {
+        type = listOf str;
+        default = [ ];
+        example = literalExpression ''
+          [
+            "--labels=Foo=Bar"
+          ]
+        '';
+        description = "Extra commandline options.";
+      };
+
+    };
+  };
+
+  config = mkIf config.services.teleport.enable {
+    environment.systemPackages = [ pkgs.teleport ];
+
+    systemd.services.teleport = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network.target" ];
+      serviceConfig = {
+        ExecStart = "${pkgs.teleport}/bin/teleport start" +
+          optionalString (length cmdlineArgs != 0) (" \\\n  " +
+            concatStringsSep " \\\n  " cmdlineArgs);
```

the args should probably be escaped so systemd won't expand %specifiers in the arguments:
```suggestion
          optionalString (length cmdlineArgs != 0) (" \\\n  " +
            concatMapStringsSep " \\\n  " builtins.toJSON cmdlineArgs);
```
`toJSON` just so happens to produce the correct escaping for systemd as well.

also: why add line continuations here? just curious.

---

a pure copy-paste from the kubelet module